src/typed/typemod.ml

```diff
@@ -284,60 +284,126 @@ let enrich_type_decls anchor decls oldenv newenv =
+    Ctype.init_def(Ident.current_time());
     let scope = None in
     let defs, newenv = Typecore.type_binding env rec_flag binds scope in
     let () = if rec_flag = Recursive then
         Typecore.check_recursive_bindings env defs
     in
-    (TTopLet(rec_flag, defs)),
-    List.map (fun id -> TSigValue(id, Env.find_value (PIdent id) newenv))
-      (let_bound_idents defs), newenv in
-
-  let init_stmts = List.map2 (fun d (_, loc) -> {ttop_desc=TTopData(d); ttop_loc=loc; ttop_env=newenv}) decls datas in
+    let some_exported = ref false in
+    let signatures = List.fold_right (fun id sigs ->
+      if ident_needs_export id || export_flag = Exported
+      then (some_exported := true; (TSigValue(id, Env.find_value (PIdent id) newenv))::sigs)
+      else sigs
+    ) (let_bound_idents defs) [] in
+    let export_flag = if !some_exported then Exported else export_flag in
+    let stmt = { ttop_desc = TTopLet(export_flag, rec_flag, defs); ttop_loc = loc; ttop_env = env } in
+    newenv, signatures, stmt in
+
+  let process_export env exports loc =
+    let bindings = List.map (fun {pex_name=name; pex_alias=alias; pex_loc=loc} ->
+      let exported_name = match alias with 
+      | Some(alias) -> alias.txt 
+      | None -> Identifier.string_of_ident name.txt in
+      let bind_name = { txt=exported_name; loc } in
+      {
+        pvb_pat={ppat_desc=PPatVar(bind_name); ppat_loc=loc};
+        pvb_expr={pexp_loc=loc; pexp_desc=PExpId name};
+        pvb_loc=loc;
+      }
+    ) exports in
+    process_let env Exported Nonrecursive bindings loc in
+
+  let process_export_data env exports loc =
+    let process_one rs {pexd_name=name; pexd_loc=loc} =
+      let type_id = Env.lookup_type (IdentName name.txt) env in
+      let type_ = Env.find_type type_id env in
+      TSigType(Path.head type_id, type_, rs) in
+    if List.length exports > 1 then
+      (process_one TRecFirst (List.hd exports))::List.map (process_one TRecNext) (List.tl exports)
+    else
+      List.map (process_one TRecNot) exports in
+
+  let final_env, signatures, statements = List.fold_left (fun (env, signatures, statements) {ptop_desc; ptop_loc=loc} ->
+      match ptop_desc with
+      | PTopImport i -> 
+        let new_env, statement = process_import env i loc in 
+        new_env, signatures, statement::statements
+      | PTopExport ex -> 
+        let new_env, sigs, statement = process_export env ex loc in 
+        new_env, List.rev sigs @ signatures, statement::statements
+      | PTopExportData ex -> 
+        let sigs = process_export_data env ex loc in 
+        env, List.rev sigs @ signatures, statements
+      | PTopForeign(e, d) -> 
+        let new_env, signature, statement = process_foreign env e d loc in
+        let signatures = match signature with Some(s) -> s::signatures | None -> signatures in
+        new_env, signatures, statement::statements
+      | PTopData(e, d) -> 
+        let new_env, sigs, statement = process_datas env e [d] loc in 
```

Comment:
One thing to point out here: the compiler was assuming that all data declarations were mutually recursive. I set this up to make it easy to parse mutually recursive datas and put them together into one `PTopData`. There will be a follow-up PR that implements that.